### ISSUE TITLE
[Bug] IO Error after frame creation without other action [sf#185]

### ISSUE THREAD

**Issue by DietmarSchwertberger** [DietmarSchwertberger/wxGlade#260]
**Reported by eanon on 2016-04-18 01:57 UTC**
Hello. I launch wxGlade, then just create a frame and I wait... Some minutes later, I get an internal error (exceptions.IOError) systematically. I'm under Windows 7 (SP1) Pro 64-bit FR with Python 2.7. I'm first talked about this issue in the wxWidgets forum during this thread: https://forums.wxwidgets.org/viewtopic.php?f=27&t=42110&p=170460

Here is a copy of the error report:

~~~~
ERROR   : Internal Error
An unexpected error occurred!

Date and time:       2016-04-18T03:46:34.282000
Python version:      2.7.3
wxPython version:    3.0.2.0
wxWidgets platform:  __WXMSW__
wxGlade version:     0.7.2
Language code:       fr_FR
Encoding:            cp1252
Filesystem encoding: mbcs

Exception type:      <type 'exceptions.IOError'>
Exception details:   [Errno 2] No such file or directory: '%HOMEDRIVE%%HOMEPATH%\\#~wxg.autosave~#'
Application stack trace:
Stack frame at level 0
======================
  File "C:\Python\lib\codecs.py", line 881
  Function "open(filename='%HOMEDRIVE%%HOMEPATH%\\#~wxg.autosave~#', mode='wb', encoding='utf-8', errors='strict', buffering=1)"
  Source code context:
              if 'b' not in mode:
                  # Force opening of the file in binary mode
                  mode = mode + 'b'
  ->      file = __builtin__.open(filename, mode, buffering)
          if encoding is None:
              return file
          info = lookup(encoding)
  Local variables:
  -> filename (<type 'str'>): '%HOMEDRIVE%%HOMEPATH%\\#~wxg.autosave~#'
  -> mode (<type 'str'>): 'wb'
  -> buffering (<type 'int'>): 1
  -> errors (<type 'str'>): 'strict'
  -> encoding (<type 'str'>): 'utf-8'

Stack frame at level 1
======================
  File "C:\Program Files (x86)\wxGlade\common.py", line 576
  Function "autosave_current()"
  Source code context:
              return False         # do nothing in this case...
          try:
              autosave_name = get_name_for_autosave()
  ->          outfile = codecs.open(autosave_name, 'w', 'utf-8')
              app_tree.write(outfile)
              outfile.close()
          except:
  Local variables:
  -> autosave_name (<type 'str'>): '%HOMEDRIVE%%HOMEPATH%\\#~wxg.autosave~#'
~~~~

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-04-18 19:04 UTC**
Hi eanon,

on Windows wxGlade use the environment variable "%USERPROFILE%" to determinate user's home directory. 

In your case the "%USERPROFILE%" is "%HOMEDRIVE%%HOMEPATH%" and it looks like both variables don't exists, even if they are mandatory.

In my systems they are set like (stripped output of the "set" command in a terminal window):
```
HOMEDRIVE=C:
HOMEPATH=\Dokumente und Einstellungen\carsten
USERNAME=carsten
USERPROFILE=C:\Dokumente und Einstellungen\carsten
```

Could you share your version of this three variables?

There is a similar topic written in this thread: http://answers.microsoft.com/en-us/windows/forum/all/homepath-variable-has-disappeared/d1d12afd-4b06-4c6f-ba16-bd189fef198f?auth=1 You may probably also have such a permission issue.

Regarding the crashed mentioned in the forum: wxGlade 0.7.2 has a bug (a kind of use after free) within the copy/cut and paste logic. Thereby wxGlade crashes on pasting widgets. This bug is fixed now. 

You can download the development version from Bitbucket (https://bitbucket.org/agriggio/wxglade), extract the archive to a directory and launch wxGlade from that directory.

Regards,
Carsten

**Comment by DietmarSchwertberger** [comment]
**Updated by cgrohmann on 2016-04-18 19:04 UTC**
- **assigned_to**: Carsten Grohmann

**Comment by DietmarSchwertberger** [comment]
**Commented by eanon on 2016-04-19 01:55 UTC**
Hello Carsten,

Here are the values of these env. variables from set command:

~~~~
HOMEDRIVE=C:
HOMEPATH=\Users\eln
USERNAME=eln
USERPROFILE=C:\Users\eln
~~~~

So, my variables are well here and it sounds I'm not in the same case as the one described in the page of the link you gave me.

Well, however, if I save the project, it seems to be OK. The crash only happened when I created a wxFrame in an "untitled" project. Otherwise said, in a very rare case (for example, if I start a frame, then the phone ring:)

Best,
Eric

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-04-19 05:21 UTC**
Hello Eric,

the error occured in the auto-save mechanism. May you deactivate auto-save within the preferences menu in the meanwhile.

Regards,
Carsten

**Comment by DietmarSchwertberger** [comment]
**Commented by eanon on 2016-04-19 15:56 UTC**
OK, Cartsen, done ! even if at this stage I didn't seen this crash again (my project being engaged; ie. named/saved on disk). If you need more info or want I do something, just tell me...

Cordially

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-04-19 20:20 UTC**
Hi Eric,

I did some more research about the environment variables HOMEDRIVE and HOMEDIR. Both will be set during the first initialisation of the explorer. They may not be set if users are access windows with a second login or if the application runs as a service. 

Can you share the wxGlade log file with me? I want to check the paths printed during the initialisation phase of wxGlade. If you want, you can send the log file via email.

Do you start wxGlade from a normal user session, via runas or RDP/VNC?

What Windows version do you use?

Regards,
Carsten

**Comment by DietmarSchwertberger** [comment]
**Commented by eanon on 2016-04-21 03:27 UTC**
Hi Carsten,

No problem to send you a wxGlade's log, but I don't know your email address (clicking on your name send me to a 403 error).

I start wxGlade locally, logged as eln which is a username in the admin group. About my Windows version, I indicated it in my first post above : Windows 7 (SP1) Pro 64-bit FR

Soon
Eric

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-04-21 19:41 UTC**
Hi Eric,

you'll find my email address at the bottom of the wxGlade homepage at http://wxglade.sf.net.

Thank you for the short notice of my inaccessible user profile. I'll fix it.

Regards,
Carsten

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-04-22 19:06 UTC**
Thank you for sharing the logs. The log file shows the failed expansion already. 
~~~
2016-04-22 15:11:46,694 root INFO: Home directory: %HOMEDRIVE%%HOMEPATH%
~~~

I'll fix it within the next few days.

Regards,
Carsten

**Comment by DietmarSchwertberger** [comment]
**Updated by cgrohmann on 2016-04-29 18:49 UTC**
- **status**: open --> closed

**Comment by DietmarSchwertberger** [comment]
**Commented by cgrohmann on 2016-04-29 18:49 UTC**
Hi Eric,

his bug is fixed in changeset 1930:5b128b190bbc "Rework querying home path to prevent unexpanded variables (sf bug #185)". The following changeset (1931:ac5e47757e72 "Disable autosave functionality permanently after saving has failed") disables the autsave functionality if an error occurred and shows an explainatory message box.

The fix will be part of the next offical release.

If you need the fix more urgent, then you may download the current development version from https://bitbucket.org/agriggio/wxglade/get/tip.tar.gz .

Thank you for reporting this bug,
Carsten

**Comment by DietmarSchwertberger** [comment]
**Commented by eanon on 2016-06-01 18:33 UTC**
Thanks, Carsten. I'm a "little" busy these days, but appreciated your smart support and will update wxGlade soon (I opted for wxGlade for my Python projects).

Cordially
Eric
